soc: ti_k3: specify soc-specific ctrl partitions

natto1784 opened the pull request:
Tried moving the code for padcfg mmr unlocks in this patch

Reasons:
1. The base addresses can be different for different socs and domains
2. It only worked for the m4 core, it should not be only limited to that. However, right now the only changes I've made is for am62_m4
3. The sys_write calls were to the wrong addresses (not translated via the RAT module)

Note: Changes only reflected for am62 here but tested for r5 and m4 standalone cores on am64x/am243x (separate patch in near future)

natto1784 commented:
CC: @vaishnavachath @glneo

vaishnavachath commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -6,31 +6,43 @@
 
+	if (padcfg_dev) {
+		k3_lock_mmr_partition(padcfg_dev, 0);
+		k3_lock_mmr_partition(padcfg_dev, 1);
```
Will this part == 0 or 1 be always true for all devices?

natto1784 replied:
It should be the case for all K3 devices, but I understand that a separate dt-binding might be in order for partition addresses

vaishnavachath commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -6,31 +6,43 @@
 #include <zephyr/drivers/pinctrl.h>
 
 #define PINCTRL_NODE DT_NODELABEL(pinctrl)
+#define PADCFG_NODE  DT_NODELABEL(padcfg)
```
could we make this implementation generic, something like consumernode_ctrlmmr, here it will be padcfg_ctrlmmr

glneo commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -6,31 +6,43 @@
-} pinctrl_ti_k3_cfg = {
-	DEVICE_MMIO_ROM_INIT(PINCTRL_NODE)
-};
+} pinctrl_ti_k3_cfg = {DEVICE_MMIO_ROM_INIT(PINCTRL_NODE)};
```
Unrelated code format change? Should remove these.

natto1784 replied:
Ah, ran clang-format on changed files, thought it would be good to leave these changes, but will remove them from now on 👍

glneo commented on drivers/pinctrl/pinctrl_ti_k3.c:
```diff
@@ -40,11 +52,5 @@ static int pinctrl_ti_k3_init(const struct device *dev)
-		 PRE_KERNEL_1,
-		 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
-		 NULL);
+DEVICE_DT_DEFINE(PINCTRL_NODE, pinctrl_ti_k3_init, NULL, &pinctrl_ti_k3_dev, &pinctrl_ti_k3_cfg,
```
Unrelated code format change..

glneo commented on soc/ti/k3/CMakeLists.txt:
```diff
@@ -1,4 +1,7 @@
 # Copyright (c) 2024 Nordic Semiconductor ASA
 # SPDX-License-Identifier: Apache-2.0
 
+zephyr_include_directories(common)
```
This can mess up things, better to be clear and use the full path in the code below,

```
#include <common/k3_mmr.h>
```

glneo commented on dts/arm/ti/am62x_m4.dtsi:
```diff
@@ -48,6 +48,12 @@
+	padcfg: padcfg@4080000 {
+		compatible = "syscon";
+		reg = <0x04080000 DT_SIZE_K(32)>;
+		status = "okay";
```
Status is default "okay", this line then isn't needed (unless you are updating an existing disabled node which is not the  case here).

natto1784 replied:
syscon uses the DT_INST_FOREACH_STATUS_OKAY macro to initialize all compatible nodes with status set as "okay"

glneo commented on dts/arm/ti/am62x_m4.dtsi:
```diff
@@ -48,6 +48,12 @@
 
+	padcfg: padcfg@4080000 {
+		compatible = "syscon";
+		reg = <0x04080000 DT_SIZE_K(32)>;
```
Hmm, looks like you are also trying to unlock the pinctrl MMR also?.. I wonder if the pinctrl driver should be doing that

natto1784 replied:
Partitions lie at a difference of 16k each. Two partitions would mean 32k. This indeed does overlap with the pinctrl regs, which can be a problem. Thinking of a different approach to this. Maybe adding a new property to k3 pinctrl can be a solution.

glneo commented on dts/arm/ti/am62x_m4.dtsi:
```diff
@@ -48,9 +48,22 @@
 	pinctrl: pinctrl@4084000 {
 		compatible = "ti,k3-pinctrl";
 		reg = <0x04084000 0x88>;
+		ctrl-partitions = <&padcfg_part0 &padcfg_part1>;
```
Why do we need a phandle to our own memory range? Could you instead call this new `k3_unlock_ctrl_partition()` for all the "ti,k3-ctrl-partition" nodes directly from `soc_early_init_hook()`?

natto1784 replied:
I am not sure if unlocking all partitions at the beginning is the correct way to go. Currently, pinctrl unlocks, writes and locks these partitions. We can indeed unlock all the partitions in `soc_early_init_hook()` and lock them again in `soc_late_init_hook()` if that is more preferrable

gramsay0 replied:
U-Boot just unlocks everything during SoC init (no device tree additions), and doesn't appear to lock it again:
https://github.com/u-boot/u-boot/blob/5061eab96acd60d28149e080f3a37fbe7220ca54/arch/arm/mach-k3/am62x/am625_init.c#L47

pinctrl can be changed at runtime, so I don't think locking during `soc_late_init_hook` is ok

natto1784 replied:
Noted, won't lock it after kernel init then. Although if the application does want to change the pinctrl at runtime, it can unlock it again, right? Albeit with one write worth of overhead. 

By the way, what are your opinions on whether we should do ifdefs for different SoCs' MMR addresses instead of using DT?

natto1784 commented:
Side note: Currently we only have one instance of pinmux for every SoC using this driver. It would be great if we could have multiple such nodes, to be able to use `main` and `mcu` domain peripherals such as uart together.

gramsay0 commented:
> Side note: Currently we only have one instance of pinmux for every SoC using this driver. It would be great if we could have multiple such nodes, to be able to use `main` and `mcu` domain peripherals such as uart together.

For reference, here was an earlier attempt to align pincrtl device tree with Linux (which uses "pinctrl-single" compat) and enable multi-instance pinctrl: https://github.com/zephyrproject-rtos/zephyr/pull/56677

The part to make it multi-instance is mostly just ("pinctrl_soc.h"):
```diff
struct pinctrl_soc_pin {
+	const struct device *dev;
	uint32_t offset;
	uint32_t value;
};

typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;

#define TI_SITARA_DT_PIN(node_id)				\
	{							\
+		.dev = DEVICE_DT_GET(DT_PARENT(node_id)),	\
```

natto1784 commented:
> > Side note: Currently we only have one instance of pinmux for every SoC using this driver. It would be great if we could have multiple such nodes, to be able to use `main` and `mcu` domain peripherals such as uart together.
> 
> For reference, here was an earlier attempt to align pincrtl device tree with Linux (which uses "pinctrl-single" compat) and enable multi-instance pinctrl: #56677
> 
> The part to make it multi-instance is mostly just ("pinctrl_soc.h"):
> 
> ```diff
> struct pinctrl_soc_pin {
> +	const struct device *dev;
> 	uint32_t offset;
> 	uint32_t value;
> };
> 
> typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
> 
> #define TI_SITARA_DT_PIN(node_id)				\
> 	{							\
> +		.dev = DEVICE_DT_GET(DT_PARENT(node_id)),	\
> ```

This is what I thought of first as well. I tried it and it works fine with `DT_INST_FOREACH_STATUS_OKAY` in `pinctrl_ti_k3.c`

Commit:
soc: ti_k3: specify soc-specific ctrl partitions

The first two partitions in padconfig mmr regions need to be unlocked for
the pinctrl driver to be able to write. However, the base addresses for the
the registers can be different across SoCs and domains. Besides, currently
this is only done for M4 and that too not in the local (M4's) view.

This patch introduces a file specifying all ctrl partition base addresses
using ifdef directives for different SoCs and variants, and unlocking them
before the kernel and drivers initialize.

Signed-off-by: Amneesh Singh <a-singh7@ti.com>

gramsay0 commented on soc/ti/k3/am6x/common/ctrl_partitions.c:
```diff
@@ -0,0 +1,42 @@
+#define KICK1_UNLOCK_VAL (0xD172BC5AU)
+#define KICK_LOCK_VAL    (0x0U)
+
+const uintptr_t ctrl_partitions[] = {
```
`static`

vaishnavachath commented on soc/ti/k3/am6x/common/ctrl_partitions.c:
```diff
@@ -0,0 +1,42 @@
+const uintptr_t ctrl_partitions[] = {
+#if defined CONFIG_SOC_AM6234_M4 || defined CONFIG_SOC_AM6442_M4
+	0x4080000, /* mcu/wkup padcfg 1 */
+	0x4084000, /* mcu/wkup padcfg 2 */
```
Instead of keeping raw addresses here, can you add a define and also derive the second page address from the first using a macro? ```base + partition * CTRL_MMR0_PARTITION_SIZE```

gramsay0 commented on soc/ti/k3/am6x/common/ctrl_partitions.c:
```diff
@@ -0,0 +1,54 @@
+#define KICK1_UNLOCK_VAL (0xD172BC5AU)
+#define KICK_LOCK_VAL    (0x0U)
+
+#define PARTITION_SIZE             (0x4000)
+#define CTRL_PARTITION(base, part) (base + part * PARTITION_SIZE)
```
```suggestion
#define CTRL_PARTITION_SIZE        (0x4000)
#define CTRL_PARTITION(base, part) ((base) + (part) * CTRL_PARTITION_SIZE)
```

gramsay0 replied:
The macro parameters should also have brackets

vaishnavachath commented on soc/ti/k3/am6x/common/ctrl_partitions.c:
```diff
@@ -0,0 +1,54 @@
+static uintptr_t translate_addr(uintptr_t phys)
+{
+#ifdef CONFIG_MM_TI_RAT
+	uintptr_t virt;
```
@natto1784 Won't you need to do this for A53 as well (today it boots with U-Boot, so its already unlocked), can you use generic DEVICE_MMIO/device_map helpers instead of sys_mm_drv_page_phys_get()?

natto1784 commented:
done @kartben

natto1784 commented:
Hi @kartben, is there anything blocking this?
